aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu

```diff
@@ -58,16 +58,69 @@ void foreach_tensor_##NAME##_scalarlist_kernel_cuda_(TensorList tensors, at::Arr
 std::vector<Tensor> foreach_tensor_##NAME##_scalarlist_kernel_cuda(TensorList tensors, at::ArrayRef<Scalar> scalars) {  \
     check_foreach_api_restrictions(tensors, scalars);                                                                   \
-    if (!can_use_fast_route(tensors, scalars)) {                                                                        \
+    if (!can_use_fast_route({tensors}, scalars)) {                                                                      \
         return at::native::foreach_tensor_##NAME##_scalarlist_kernel_slow(tensors, scalars);                            \
     }                                                                                                                   \
                                                                                                                         \
     return foreach_binary_op<OP>(tensors, scalars);                                                                     \
 }
 
 FOREACH_BINARY_OP_SCALARLIST(add, std::plus);
-FOREACH_BINARY_OP_SCALARLIST(sub, std::minus);
 FOREACH_BINARY_OP_SCALARLIST(mul, std::multiplies);
-FOREACH_BINARY_OP_SCALARLIST(div, std::divides);
+
+// In the case of division, integer inputs will result in float. 
+// Currently multi tensor apply can only return result of the same type as input.
+void foreach_tensor_div_scalarlist_kernel_cuda_(TensorList tensors, at::ArrayRef<Scalar> scalars) {
+    check_foreach_api_restrictions(tensors, scalars);
+    if (!can_use_fast_route(tensors, scalars, /*div_op*/ true)) {
+        return at::native::foreach_tensor_div_scalarlist_kernel_slow_(tensors, scalars);
+    }
+
+    foreach_binary_op_<std::divides>(tensors, scalars);
+}
+
+std::vector<Tensor> foreach_tensor_div_scalarlist_kernel_cuda(TensorList tensors, at::ArrayRef<Scalar> scalars) {
+    check_foreach_api_restrictions(tensors, scalars);
+    if (!can_use_fast_route(tensors, scalars, /*div_op*/ true)) {
+        return at::native::foreach_tensor_div_scalarlist_kernel_slow(tensors, scalars);
+    }
+
+    return foreach_binary_op<std::divides>(tensors, scalars);
+}
+
+// In the case of subtraction, we dont allow scalar to be boolean following the torch.sub logic
+void foreach_tensor_sub_scalarlist_kernel_cuda_(TensorList tensors, at::ArrayRef<Scalar> scalars) {
+    check_foreach_api_restrictions(tensors, scalars);
+
+    TORCH_CHECK(tensors[0].scalar_type() != kBool || !scalars[0].isBoolean(),
+              "Subtraction, the `-` operator, with two bool tensors is not supported. "
+              "Use the `^` or `logical_xor()` operator instead.")
+    TORCH_CHECK(tensors[0].scalar_type() != kBool && !scalars[0].isBoolean(),
+              "Subtraction, the `-` operator, with a bool tensor is not supported. "
+              "If you are trying to invert a mask, use the `~` or `logical_not()` operator instead.");
```

Comment:
The error messages are more maintainable if we toss them into the same check. It would be nice to unify these error messages with those in BinaryOps.h as well